### PR TITLE
msg/async, v2: authenticated onwire encryption and new preamble format

### PR DESCRIPTION
**Pull request by rzarzynski**

### CODE REVIEW

File: src/msg/async/crypto_onwire.cc [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+
+void AES128GCM_OnWireTxHandler::authenticated_encrypt_update(
+  ceph::bufferlist&& plaintext)
+{
+  int update_len = 0;
+  auto filler = buffer.append_hole(plaintext.length());
+  if(1 != EVP_EncryptUpdate(ectx.get(),
+	reinterpret_cast<unsigned char*>(filler.c_str()),
+	&update_len,
+	reinterpret_cast<const unsigned char*>(plaintext.c_str()),
```
Comment (liewegas) [inline src/msg/async/crypto_onwire.cc]: we may eventually want to avoid c_str() on plaintext to allow a gatherlist?

Comment (rzarzynski) [reply src/msg/async/crypto_onwire.cc]: Definitely TODO.

File: src/msg/async/ProtocolV2.cc [hunk 2]
```diff
@@ -1074,123 +1186,10 @@ bool ProtocolV2::is_queued() {
...
-    in.push_back(buffer::create_static(length, payload));
-    bufferlist out;
-    if (session_security->decrypt_bufferlist(in, out) < 0) {
-      throw DecryptionError();
-    }
-    ceph_assert(out.length() <= length);
-    memcpy(payload, out.c_str(), out.length());
-    length = out.length();
-  }
+void ProtocolV2::authencrypt_payload(bufferlist &payload) {
```
Comment (rjfd) [inline src/msg/async/ProtocolV2.cc]: We can remove this function, right?

File: src/msg/async/ProtocolV2.cc [hunk 3]
```diff
@@ -1074,123 +1186,10 @@ bool ProtocolV2::is_queued() {
...
-    *enc_pad_len = enc_pad_l;
-  }
-  if (total_len) {
-    *total_len = total_l;
-  }
-
-  ldout(cct, 21) << __func__ << " length=" << length << " total_len=" << total_l
-                 << " sig_pad_len=" << sig_pad_l << " enc_pad_len=" << enc_pad_l
-                 << dendl;
+void ProtocolV2::authdecrypt_payload(char *payload, uint32_t &length) {
```
Comment (rjfd) [inline src/msg/async/ProtocolV2.cc]: Same here (remove it?)

File: src/msg/async/ProtocolV2.cc [hunk 4]
```diff
@@ -117,28 +88,114 @@ static void alloc_aligned_buffer(bufferlist &data, unsigned len, unsigned off) {
  * Protocol V2 Frame Structures
  **/
 
-// this type is used to specify the position of the signature within a frame
-struct signature_t {};
+static constexpr uint8_t CRYPTO_BLOCK_SIZE { 16 };
+
+using segment_t = ProtocolV2::segment_t;
+
+// V2 preamble is consisted with one or many preable blocks depending
```
Comment (rjfd) [inline src/msg/async/ProtocolV2.cc]: nit: fixing some typos:
```
// V2 preamble consists of one or more preamble blocks depending
// on the number of segments a particular frame needs. Each block holds
// up to 2 segments and has its own CRC.
```

File: src/msg/async/ProtocolV2.cc [hunk 5]
```diff
@@ -117,28 +88,114 @@ static void alloc_aligned_buffer(bufferlist &data, unsigned len, unsigned off) {
...
+  __u8 tag;
+
+  // Number of segments to go in entire frame. First preable block has
+  // set this to just #segments, second #segments - MAX_NUM_SEGMENTS,
+  // third to #segments - MAX_NUM_SEGMENTS and so on.
+  __u8 num_segments;
+
+  std::array<ProtocolV2::segment_t, MAX_NUM_SEGMENTS> segments;
+
+  uint8_t num_extra_preamble_blocks() const {
```
Comment (rjfd) [inline src/msg/async/ProtocolV2.cc]: This method is not used anywhere. Should we remove it, or will it be used when implementing the support for reading multiple preamble blocks?

Comment (rzarzynski) [reply src/msg/async/ProtocolV2.cc]: Yeah, it was for preamble chaining. We still have possibility to start doing that in the future but, at the moment, it's unnecessary as preamble became 32 bytes long. To be dropped.

File: src/msg/async/ProtocolV2.cc (hunk 4 above)
Comment (rzarzynski) [reply src/msg/async/ProtocolV2.cc]: Thanks!

File: src/msg/async/ProtocolV2.cc (hunk 2 above)
Comment (rzarzynski) [reply src/msg/async/ProtocolV2.cc]: TODO.

File: src/msg/async/crypto_onwire.cc [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+	plaintext.length())) {
+    throw std::runtime_error("EVP_EncryptUpdate failed");
```
Comment (rzarzynski) [inline src/msg/async/crypto_onwire.cc]: TODO/WIP: `_fault()` on exception in `Protocol2.cc`.

File: src/msg/async/crypto_onwire.cc [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+
+ceph::bufferlist AES128GCM_OnWireRxHandler::authenticated_decrypt_update(
+  ceph::bufferlist&& ciphertext,
+  std::uint32_t alignment)
+{
+  ceph_assert(ciphertext.length() > 0);
+  //ceph_assert(ciphertext.length() % AESGCM_BLOCK_LEN == 0);
+
+  // TODO: consider in-place transformata
+  //auto plaintext = ceph::buffer::ptr_node::create_aligned(
```
Comment (rzarzynski) [inline src/msg/async/crypto_onwire.cc]: TODO/WIP.

File: src/msg/async/crypto_onwire.cc [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+  const AuthConnectionMeta& auth_meta,
+  bool crossed)
+{
+  if (auth_meta.is_mode_secure()) {
+    // CLEANME, CLEANME CLEANME
+    ceph_assert_always(
+      auth_meta.connection_secret.length() >= 16 + 2*sizeof(ceph::uint128_t));
+
+    ceph::uint128_t rx_nonce;
+    ::memcpy(&rx_nonce, auth_meta.connection_secret.c_str() + 16, sizeof(rx_nonce));
```
Comment (rzarzynski) [inline src/msg/async/crypto_onwire.cc]: TODO/WIP: eradicate magic enirely (also applies to `ProtocolV2.{cc,h}`).

File: src/msg/async/crypto_onwire.h [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+  // called and another round of update-...-update-final will take place.
+  //
+  // The input parameter informs implementation how the -update sequence
+  // is fragmented and allows to make concious decision about allocation
+  // or reusage of provided memory. One implementation could do in-place
+  // encryption while other might prefer one huge output buffer.
+  //
+  // It's undefined what will happen if client doesn't follow the order.
+  //
+  // TODO: switch to always_aligned_t
```
Comment (rzarzynski) [inline src/msg/async/crypto_onwire.h]: TODO: block size alignment. GCM can live (and lives) without it as it's a stream cipher but ultimately we don't want to be restricted to stream ciphers only.

**Comment by liewegas** [comment]
```
   -76> 2019-02-17 17:52:05.181 7f442e9bc700 10 -- [v2:172.21.15.104:3300/0,v1:172.21.15.104:6789/0] dispatch_throttle_release 55 to dispatch throttler 32/104857600
   -75> 2019-02-17 17:52:05.183 7f442e9bc700 -1 /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.0.1-3869-g1405567/rpm/el7/BUILD/ceph-14.0.1-3869-g1405567/src/common/Throttle.cc: In function 'virtual int64_t Thr
ottle::put(int64_t)' thread 7f442e9bc700 time 2019-02-17 17:52:05.182104
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.0.1-3869-g1405567/rpm/el7/BUILD/ceph-14.0.1-3869-g1405567/src/common/Throttle.cc: 225: FAILED ceph_assert(count >= c)

 ceph version 14.0.1-3869-g1405567 (1405567b10ef50ad2e6d97fcd8b0e801204841b2) nautilus (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x14a) [0x7f4438958336]
 2: (ceph::__ceph_assertf_fail(char const*, char const*, int, char const*, char const*, ...)+0) [0x7f4438958504]
 3: (Throttle::put(long)+0x3e9) [0x7f44389df339]
 4: (DispatchQueue::dispatch_throttle_release(unsigned long)+0x41) [0x7f4438b44b81]
```

**Comment by rzarzynski** [comment]
@liewegas: it looks `dispatch_queue->dispatch_throttle_release()` was operating on not-yet-parsed `current_header` instead of `rx_segments_todo_rev`. Pushed a fixup.

File: src/include/inline_memory.h [hunk 10]
```diff
@@ -70,7 +70,10 @@ void *maybe_inline_memcpy(void *dest, const void *src, size_t l,
 
 #if defined(__GNUC__) && defined(__x86_64__)
 
+namespace ceph {
 typedef unsigned uint128_t __attribute__ ((mode (TI)));
+}
+using ceph::uint128_t;
```
Comment (rzarzynski) [inline src/include/inline_memory.h]: FIXME: it looks ARM lacks that.

File: src/msg/async/crypto_onwire.cc (hunk 1 above)
Comment (rzarzynski) [reply src/msg/async/crypto_onwire.cc]: Fixed.

File: src/msg/async/ProtocolV2.cc (hunk 4 above)
Comment (rzarzynski) [reply src/msg/async/ProtocolV2.cc]: Updated.

File: src/msg/async/ProtocolV2.cc (hunk 2 above)
Comment (rzarzynski) [reply src/msg/async/ProtocolV2.cc]: Removed.

File: src/msg/async/ProtocolV2.cc (hunk 3 above)
Comment (rzarzynski) [reply src/msg/async/ProtocolV2.cc]: Removed as well.

File: src/msg/async/ProtocolV2.cc (hunk 5 above)
Comment (rzarzynski) [reply src/msg/async/ProtocolV2.cc]: Dropped.

**Pull request merged**